quickwit/quickwit-ingest/src/ingest_v2/models.rs

```diff
@@ -17,190 +17,169 @@
-    /// Position of the last record written in the shard's mrecordlog queue.
-    pub replication_position_inclusive: Position,
-    pub new_records_tx: watch::Sender<()>,
-    pub new_records_rx: watch::Receiver<()>,
-}
-
-impl fmt::Debug for ReplicaShard {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("ReplicaShard")
-            .field("leader_id", &self.leader_id)
-            .field("shard_state", &self.shard_state)
-            .finish()
-    }
-}
 
-impl ReplicaShard {
-    pub fn new(leader_id: NodeId) -> Self {
+    pub fn new_replica(
+        leader_id: NodeId,
+        shard_state: ShardState,
+        replication_position_inclusive: Position,
+        indexing_status: IndexingStatus,
+    ) -> Self {
         let (new_records_tx, new_records_rx) = watch::channel(());
         Self {
-            leader_id,
-            shard_state: ShardState::Open,
-            replication_position_inclusive: Position::Beginning,
+            shard_type: IngesterShardType::Replica { leader_id },
+            shard_state,
+            replication_position_inclusive,
+            indexing_status,
             new_records_tx,
             new_records_rx,
         }
     }
-}
 
-/// A shard hosted on a single node when the replication factor is set to 1. When a shard is
-/// recovered after a node failure, it is always recreated as a solo shard in closed state.
-pub(super) struct SoloShard {
-    pub shard_state: ShardState,
-    /// Position of the last record written in the shard's mrecordlog queue.
-    pub replication_position_inclusive: Position,
-    pub new_records_tx: watch::Sender<()>,
-    pub new_records_rx: watch::Receiver<()>,
-}
-
-impl fmt::Debug for SoloShard {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("SoloShard")
-            .field("shard_state", &self.shard_state)
-            .finish()
-    }
-}
-
-impl SoloShard {
-    pub fn new(shard_state: ShardState, replication_position_inclusive: Position) -> Self {
+    pub fn new_solo(
+        shard_state: ShardState,
+        replication_position_inclusive: Position,
+        indexing_status: IndexingStatus,
+    ) -> Self {
         let (new_records_tx, new_records_rx) = watch::channel(());
         Self {
+            shard_type: IngesterShardType::Solo,
             shard_state,
             replication_position_inclusive,
+            indexing_status,
             new_records_tx,
             new_records_rx,
         }
     }
-}
 
-#[derive(Debug)]
-pub(super) enum IngesterShard {
-    /// A primary shard hosted on a leader and replicated on a follower.
-    Primary(PrimaryShard),
-    /// A replica shard hosted on a follower.
-    Replica(ReplicaShard),
-    /// A shard hosted on a single node when the replication factor is set to 1.
-    Solo(SoloShard),
-}
+    pub fn notify_new_records(&mut self) {
+        self.new_records_tx
+            .send(())
+            .expect("channel should be open");
```

Comment:
```suggestion
            // `new_records_tx` is guaranteed to have a received since one of them is held by `self`.
            .expect("channel should be open");
```